### PR TITLE
added transformers_config for passing arguments to the transformer

### PR DESCRIPTION
**Pull request by KennethEnevoldsen**
Added transformers_config to allow the user to pass arguments to the transformers forward pass. Most notably the output_attentions.

for convenience, I used this example to test the code:
```python
import spacy

nlp = spacy.blank("en")

# Construction via add_pipe with custom config
config = {
    "model": {
        "@architectures": "spacy-transformers.TransformerModel.v1",
        "name": "bert-base-uncased",
        "transformers_config": {"output_attentions": True},
    }
}
transformer = nlp.add_pipe(
    "transformer", config=config)
transformer.model.initialize()


doc = nlp("This is a sentence.")
```
which gives you:
```python
len(doc._.trf_data.attention) # 12
doc._.trf_data.attention[-1].shape  # (1, 12, 7, 7) last layer of attention 
len(doc._.trf_data.tensors) # 2 
doc._.trf_data.tensors[0].shape # (1, 7, 768) <-- wordpiece embedding
doc._.trf_data.tensors[1].shape  # (1, 768) <-- assuming this is the pooled embedding?
```

Sidenote: it took me quite a while to find the default config str. It might be ideal to make this into a standalone file and load it in?

### CODE REVIEW

File: spacy_transformers/architectures.py [hunk 1]
```diff
@@ -47,6 +47,7 @@ def transformer_tok2vec_v1(
     name: str,
     get_spans,
     tokenizer_config,
+    transformers_config,
```
Comment (svlandeg) [inline spacy_transformers/architectures.py:50]: As a quick note, if we want to add this parameter, it's probably best to bump the architecture version to `spacy-transformers.Tok2VecTransformer.v2` and update the documentation.

Comment (KennethEnevoldsen) [reply spacy_transformers/architectures.py:50]: Sure thing. I will take a look at it this or the next week.

Comment (svlandeg) [reply spacy_transformers/architectures.py:50]: No pressure, this is just a maintenance detail!

Comment (KennethEnevoldsen) [reply spacy_transformers/architectures.py]: Not a problem, I have committed a new change. Hope it solves it. I considered also adding `spacy-transformers.TransformerModel.v1` unsure however if you want to split it up too much.

I assume the documentation updated depending on the docstring (i.e. I don't need to add documentation elsewhere?)

File: spacy_transformers/layers/transformer_model.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
 
 
 def TransformerModel(
-    name: str, get_spans: Callable, tokenizer_config: dict
+    name: str, get_spans: Callable, tokenizer_config: dict, transformers_config: dict
```
Comment (honnibal) [inline spacy_transformers/layers/transformer_model.py:18]: @svlandeg Does it help us if we default this parameter to an empty dict? It seems very reasonable to default this to empty, and it helps with the backwards compatibility, right?

**Comment by honnibal** [comment]
@KennethEnevoldsen Thanks for this, and sorry for the delay reviewing it. I think it'll be a really nice feature, it's just a question of getting these details about the backwards compat right.

File: spacy_transformers/architectures.py (hunk 1 above)
Comment (svlandeg) [reply spacy_transformers/architectures.py]: Ah, the documentation is over at the `spacy` repo, but we can take care of that after the merge.

File: spacy_transformers/layers/transformer_model.py (hunk 2 above)
Comment (svlandeg) [reply spacy_transformers/layers/transformer_model.py:18]: Yep, agreed!

File: spacy_transformers/data_classes.py [hunk 3]
```diff
@@ -155,11 +155,13 @@ class TransformerData:
     wordpieces: WordpieceBatch
     tensors: List[FloatsXd]
     align: Ragged
+    attention: Optional[tuple] = None  # wanted to make this Optional[Tuple[FloatsXd]],
+    # but got an error I couldn't solve
```
Comment (svlandeg) [inline spacy_transformers/data_classes.py:159]: Perhaps we need to review/revisit this as well before merging. What was the error you got?

Comment (KennethEnevoldsen) [reply spacy_transformers/data_classes.py]: I just retried it and ran tests - It works without issue. It is probably from the beginning where I tried to figure the whole thing out. Sadly only recall that I spent a lot of time on that type hint 😅

**Comment by KennethEnevoldsen** [comment]
Thanks, @honnibal, and completely understandable. It is potentially a feature that could be integrated nicely with displacy for displaying the attention given to each word. Not entirely sure though seems like the judge is still out on how interpretability these attention weights are.

File: spacy_transformers/data_classes.py (hunk 3 above)
Comment (honnibal) [reply spacy_transformers/data_classes.py]: I think it needs to be `Optional[Tuple[FloatsXd, ...]]`. If you only write `Tuple[FloatsXd]`, that means a tuple of length 1, whose item is `FloatsXd`.

(Yes this syntax is cursed.)

Comment (KennethEnevoldsen) [reply spacy_transformers/data_classes.py]: Ahh yes. It should be fixed now

File: spacy_transformers/layers/transformer_model.py (hunk 2 above)
Comment (KennethEnevoldsen) [reply spacy_transformers/layers/transformer_model.py]: Added this as well while I was at it. One point is that remove transformer_tok2vec_v2 entirely (would require setting transformer_config as the last element and adding only default to it.)

I assumed this wasn't desirable and kept the two versions.

File: spacy_transformers/layers/transformer_model.py [hunk 4]
```diff
@@ -25,6 +25,7 @@ def TransformerModel(
         overlap, and you can also omit sections of the Doc if they are not
         relevant.
     tokenizer_config (dict): Settings to pass to the transformers tokenizer.
+    transformers_config (dict): Settings to pass to the transformers forward pass.
```
Comment (svlandeg) [inline spacy_transformers/layers/transformer_model.py:28]:
```suggestion
    transformer_config (dict): Settings to pass to the transformers forward pass.
```

File: spacy_transformers/layers/transformer_model.py [hunk 5]
```diff
@@ -38,6 +39,7 @@ def TransformerModel(
             "get_spans": get_spans,
             "name": name,
             "tokenizer_config": tokenizer_config,
+            "transformers_config": transformer_config,
```
Comment (svlandeg) [inline spacy_transformers/layers/transformer_model.py:42]:
```suggestion
            "transformer_config": transformer_config,
```

File: spacy_transformers/layers/transformer_model.py [hunk 6]
```diff
@@ -75,7 +77,8 @@ def init(model: Model, X=None, Y=None):
         return
     name = model.attrs["name"]
     tok_cfg = model.attrs["tokenizer_config"]
-    tokenizer, transformer = huggingface_from_pretrained(name, tok_cfg)
+    trf_cfg = model.attrs["transformers_config"]
```
Comment (svlandeg) [inline spacy_transformers/layers/transformer_model.py:80]:
```suggestion
    trf_cfg = model.attrs["transformer_config"]
```

File: spacy_transformers/layers/transformer_model.py [hunk 7]
```diff
@@ -118,6 +118,7 @@ def forward(
 ) -> Tuple[FullTransformerBatch, Callable]:
     tokenizer = model.attrs["tokenizer"]
     get_spans = model.attrs["get_spans"]
+    trf_config = model.attrs["transformers_config"]
```
Comment (svlandeg) [inline spacy_transformers/layers/transformer_model.py:121]:
```suggestion
    trf_config = model.attrs["transformer_config"]
```

File: spacy_transformers/pipeline_component.py [hunk 8]
```diff
@@ -387,7 +394,7 @@ def from_disk(
         def load_model(p):
             p = Path(p).absolute()
             tokenizer, transformer = huggingface_from_pretrained(
-                p, self.model.attrs["tokenizer_config"]
+                p, self.model.attrs["tokenizer_config"], self.model.attrs["transformers_config"]
```
Comment (svlandeg) [inline spacy_transformers/pipeline_component.py:397]:
```suggestion
                p, self.model.attrs["tokenizer_config"], self.model.attrs["transformer_config"]
```

File: spacy_transformers/tests/util.py [hunk 9]
```diff
@@ -132,6 +136,7 @@ def DummyTransformer(
             "tokenizer": DummyTokenizer(),
             "grad_factor": 1.0,
             "flush_cache_chance": 0.0,
+            "transformers_config": {"output_attentions": False}
```
Comment (svlandeg) [inline spacy_transformers/tests/util.py:139]:
```suggestion
            "transformer_config": {"output_attentions": False}
```

Comment (svlandeg) [reply spacy_transformers/tests/util.py:139]: Is this additional argument actually needed - does it not work with the default empty `dict`?

**Comment by svlandeg** [comment]
It looks like a binary file ".DS_Store" got committed by accident...

**Comment by svlandeg** [comment]
@KennethEnevoldsen : Apologies again for the late follow-up! I just did one final round of review and think we should be able to wrap this up soon :-)

**Comment by KennethEnevoldsen** [comment]
No problem will follow up on this during the weekend

File: spacy_transformers/tests/util.py (hunk 9 above)
Comment (KennethEnevoldsen) [reply spacy_transformers/tests/util.py:139]: It needs the output_attentions in line 146 where it handles the new output format.
```
    if trf_config["output_attentions"] is True:
```
but have changed that to:
```
if ("output_attentions" in trf_config) and (trf_config["output_attentions"] is True):
```

as I expect there might be some backward compatibility issues with the first approach.

**Comment by svlandeg** [comment]
Thanks again for your contribution and patience! I think this looks good - just made an internal note about the documentation & moving v1 to legacy, we'll take care of that in a separate PR.

**Pull request merged**